src/capacity.rs

```diff
@@ -0,0 +1,164 @@
+pub struct AudioRenderCapacity {
+    sender: Sender<LoadValueData>,
+    receiver: Option<Receiver<LoadValueData>>,
+    callback: Arc<Mutex<Option<EventHandler>>>,
+}
+
+impl AudioRenderCapacity {
+    pub(crate) fn new(sender: Sender<LoadValueData>, receiver: Receiver<LoadValueData>) -> Self {
+        let callback = Arc::new(Mutex::new(None));
+
+        Self {
+            sender,
+            receiver: Some(receiver),
+            callback,
+        }
+    }
+
+    /// Start metric collection and analysis
+    #[allow(clippy::missing_panics_doc)]
+    pub fn start(&mut self, options: AudioRenderCapacityOptions) {
+        let receiver = match self.receiver.take() {
+            None => return,
+            Some(receiver) => receiver,
+        };
+
+        let callback = self.callback.clone();
+
+        let mut timestamp: f64 = 0.;
+        let mut load_sum: f64 = 0.;
+        let mut counter = 0;
+        let mut peak_load: f64 = 0.;
+        let mut underrun_sum = 0;
+
+        let mut next_checkpoint = options.update_interval;
+        std::thread::spawn(move || {
+            for item in receiver {
+                let LoadValueData {
+                    render_timestamp,
+                    load_value,
+                } = item;
+
+                // check for signal to stop
+                if render_timestamp.is_nan() && load_value.is_nan() {
+                    return; // stop thread
+                };
+
+                counter += 1;
+                load_sum += load_value;
+                peak_load = peak_load.max(load_value);
+                if load_value > 1. {
+                    underrun_sum += 1;
+                }
+
+                if render_timestamp >= next_checkpoint {
+                    let event = AudioRenderCapacityEvent::new(
+                        timestamp,
+                        load_sum / counter as f64,
+                        peak_load,
+                        underrun_sum as f64 / counter as f64,
+                    );
+                    if let Some(f) = &mut *callback.lock().unwrap() {
+                        (f)(event);
+                    }
+
+                    next_checkpoint += options.update_interval;
+                    timestamp = render_timestamp;
+                    load_sum = 0.;
+                    counter = 0;
+                    peak_load = 0.;
+                    underrun_sum = 0;
+                }
+            }
+        });
+    }
+
+    /// Stop metric collection and analysis
+    pub fn stop(self) {
+        // halt callback thread
+        let signal = LoadValueData {
+            render_timestamp: f64::NAN,
+            load_value: f64::NAN,
+        };
+        let _ = self.sender.send(signal);
+    }
+
+    /// An EventHandler for [`AudioRenderCapacityEvent`].
+    #[allow(clippy::missing_panics_doc)]
+    pub fn onupdate<F: FnMut(AudioRenderCapacityEvent) + Send + 'static>(&mut self, callback: F) {
```

Comment:
This may collide with your work on #219 so let's make sure it aligns later

Reply:
Actually, I didn't went further as for now, I'm a bit stuck in big refactoring issues with no mental space left :)... but I think the API I ended up was pretty much the same, really cool to have this example